scheduler/data-flow/src/main/kotlin/io/seldon/dataflow/PipelineSubscriber.kt

```diff
@@ -141,12 +141,57 @@ class PipelineSubscriber(
         kafkaConsumerGroupIdPrefix: String,
         namespace: String,
     ) {
-        logger.info(
-            "Create pipeline {pipelineName}  version: {pipelineVersion} id: {pipelineId}",
-            metadata.name,
-            metadata.version,
-            metadata.id,
-        )
+        // If a pipeline with the same id exists, we assume it has the same name & version
+        // If it's in an error state, try re-creating.
+        //
+        // WARNING: at the moment handleCreate is called sequentially on each update in
+        // Flow<PipelineUpdateMessage> from subscribePipelines(). This allows us to sidestep issues
+        // related to race conditions on `pipelines.containsKey(...)` below. If we ever move to
+        // concurrent creation of pipelines, this needs to be revisited.
+        if (pipelines.containsKey(metadata.id)) {
+            val previous = pipelines[metadata.id]!!
+            if (!previous.status.isError) {
+                client.pipelineUpdateEvent(
+                    makePipelineUpdateEvent(
+                        metadata = metadata,
+                        operation = PipelineOperation.Create,
+                        success = true,
+                        reason = previous.status.getDescription() ?: "pipeline created",
```

Comment:
given that this code is used also elsewhere (line 255 in this file) should we at least make `pipeline created` message as a constant?

Reply:
also this can be in a rebalance state / creating state? ie if the dataflow in loading the pipeline, the scheduler restarted and resending the load request to the same pipeline.

or the assumption as you mentioned that it is single threaded so we guarantee that the first load cmd has finished (it is not clear 100% for me).

Reply:
Thanks for the detailed response, what happens if the scheduler disconnects while dataflow is processing `handleCreate`?

Reply:
The results of tests here indicate reasonable behaviour: if the scheduler goes away while dataflow is in `handleCreate`, the pipeline creation logic continues to work correctly (the pipeline gets created, waited for to get into "Running" state, etc). Now, we have two cases, both ending up with the similar behaviour:

1. If the scheduler is not yet up when `handleCreate` calls the first grpc function (to update the scheduler state)
 - a io.grpc.StatusException is thrown from `handleCreate()`, which is caught in `subscribePipelines()`, which finishes by reporting the error
 - the functionality to retry connecting to the scheduler in `subscribe()` kicks in, and continues until a scheduler becomes available
 - when the new scheduler becomes available, dataflow connects to it, receives new "Create" messages, and responds with the actual status of the pipelines (which we failed to transmit initially). If the pipeline ended up in an error state, we retry creating it at this point
2. If a new scheduler instance has come up in the meantime
- the io.grpc.StatusException still gets thrown, because the previous grpc connection transport has dropped; a new connection is established via the retry mechanism in `subscribe()` and everything behaves as in case 1. above.

A note on how this is tested:
 - if one restarts dataflow-engine, the scheduler re-sends pipeline create messages to the new instance, therefore `handleCreate()` gets called. However, the creation process takes a long time (tens of seconds), because Kafka Streams needs to sync with remote state from Kafka topics. This gives us time to kill the scheduler while `handleCreate()` waits for the pipeline to start and transition to a `Running` state.